jobbergate-api/jobbergate_api/s3_manager.py

```diff
@@ -1,37 +1,157 @@
 """
 Provides a convenience class for managing calls to S3.
 """
+import tarfile
+import typing
+from abc import ABC, abstractmethod
+from io import BytesIO
 
 import boto3
+from botocore.exceptions import BotoCoreError
 from fastapi import HTTPException, UploadFile, status
+from loguru import logger
 
 from jobbergate_api.config import settings
 
 
-class S3Manager:
+class BucketClientBase(ABC):
+    """
+    Base class used to describe how to interact with bucket clients.
     """
-    Provide a class for managing connection to an S3 bucket.
 
-    Prepares the bucket_name and s3 key template from project settings. Also initialize an s3
-    client. Provide methods as thin wrappers around s3 calls but provide the mappings for buckets and keys.
+    @abstractmethod
+    def put(self, file: typing.IO, key: str) -> None:
+        """
+        Upload a file to the client for the given key.
+        """
+        pass
+
+    @abstractmethod
+    def delete(self, key: str) -> None:
+        """
+        Delete a file from the client associated to the given key.
+        Raise KeyError if the file does not exist.
+        """
+        pass
+
+    @abstractmethod
+    def get(self, key: str) -> typing.IO:
+        """
+        Get a file from the client associated to the given key.
+        Raise KeyError if the file does not exist.
+        """
+        pass
+
+
+class DummyClient(BucketClientBase):
```

Comment:
I'm not sure why this is necessary. Why not just mock the s3 client in the tests?

Reply:
For me, `DummyClient` makes testing and debugging easier, and test files are also a bit cleaner without the mocks. I'll keep this suggestion in mind while solving the other ones.

Reply:
> Question: will the job scripts be in the same bucket as the applications?

They are for now. But we can change it if necessary. Do you think separated buckets are better?

Reply:
Good points around `DummyClient`, it was removed in [75b4341](https://github.com/omnivector-solutions/jobbergate/pull/111/commits/75b434114eb0bbfac0917da9c87101bb9c9eae79)